### PR TITLE
NVIDIA 2D Thread Scheduler Fixed

### PR DESCRIPTION
**Pull request by jjfumero**
#### Description

This patch provides a new Thread Scheduler for NVIDIA GPUs. 

#### Problem description

The problem is that, when using the latest NVIDIA Drivers (e.g., 550.76), the thread block is set to 32x32 for 2D kernels. This block size seems to be illegal only when using the latest NVIDIA drivers. This patch provides a custom NVIDIA scheduler to fix this. Performance over the default scheduler increases ~300GFLOPs on my RTX 3070 GPU for the canonical matrix multiplications with this patch. 

#### Backend/s tested

Mark the backends affected by this PR.

- [X] OpenCL
- [ ] PTX
- [ ] SPIRV

#### OS tested

Mark the OS where this PR is tested.

- [X] Linux
- [ ] OSx
- [ ] Windows

#### Did you check on FPGAs?

If it is applicable, check your changes on FPGAs.

- [ ] Yes
- [X] No

#### How to test the new patch?

```bash
make BACKEND=opencl
tornado --threadInfo --jvm="-Ds0.t0.device=0:0" -m tornado.examples/uk.ac.manchester.tornado.examples.compute.MatrixMultiplication2D 512
tornado --threadInfo --jvm="-Ds0.t0.device=0:0" -m tornado.examples/uk.ac.manchester.tornado.examples.compute.MatrixMultiplication2D 1024
```

### CODE REVIEW

**Comment by jjfumero** [comment]
Related issue: https://github.com/beehive-lab/TornadoVM/pull/356

File: tornado-drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/TornadoPlatformInterface.java [hunk 1]
```diff
@@ -2,7 +2,7 @@
  * This file is part of Tornado: A heterogeneous programming framework:
  * https://github.com/beehive-lab/tornadovm
  *
- * Copyright (c) 2020, APT Group, Department of Computer Science,
+ * Copyright (c) 2020, 2024 APT Group, Department of Computer Science,
```
Comment (stratika) [inline tornado-drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/TornadoPlatformInterface.java:5]:
```suggestion
 * Copyright (c) 2020, 2024, APT Group, Department of Computer Science,
```

File: tornado-drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/scheduler/OCLNVIDIAGPUScheduler.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            value--;
+        }
+        if (value >= 32 && dim > 1) {
+            value /= 2;
+        }
+        return value;
+    }
+
+    private long[] calculateEffectiveMaxWorkItemSizes(TaskMetaData metaData) {
+        long[] intermediates = new long[] { 1, 1, 1 };
```
Comment (stratika) [inline tornado-drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/scheduler/OCLNVIDIAGPUScheduler.java:96]: any particular reasoning for this particular name of the array? Shouldn't this reflect the max work items?

Comment (jjfumero) [reply tornado-drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/scheduler/OCLNVIDIAGPUScheduler.java]: Yes, I agree. I just did a refactoring of this method

**Review by stratika — APPROVED** [review]
LGTM

**Pull request merged**